### PR TITLE
Fix ExpressionEquivalence for special forms involving LIKE

### PR DESCRIPTION
**Pull request by wendigo**
Fixes https://github.com/trinodb/trino/issues/21235

### CODE REVIEW

**Comment by martint** [comment]
Can you add a test?

**Comment by wendigo** [comment]
@martint is this enough?

**Comment by findepi** [comment]
I think we should remove https://github.com/trinodb/trino/blob/278a2ab25a6e56a5d40b671aca41c4408fe6444c/core/trino-main/src/main/java/io/trino/sql/planner/optimizations/ExpressionEquivalence.java#L265-L267  (replace with throwing)
This is incorrect and that's exactly what `Comparison method violates its general contract!` message complains about.

File: core/trino-main/src/main/java/io/trino/type/LikePattern.java [hunk 1]
```diff
@@ -90,4 +91,10 @@ public String toString()
                 .add("escape", escape)
                 .toString();
     }
+
+    @Override
+    public int compareTo(LikePattern other)
+    {
+        return pattern.compareTo(other.pattern);
```
Comment (findepi) [inline core/trino-main/src/main/java/io/trino/type/LikePattern.java:98]: compare escape as well to make this linear ordering

File: core/trino-main/src/main/java/io/trino/sql/planner/optimizations/ExpressionEquivalence.java [hunk 2]
```diff
@@ -262,9 +263,7 @@ public int compare(RowExpression left, RowExpression right)
                     }
                 }
 
-                // value is some random type (say regex), so we just randomly choose a greater value
-                // todo: support all known type
-                return -1;
+                throw new IllegalArgumentException("Cannot compare " + leftConstant + " and " + rightConstant + " of type " + javaType);
```
Comment (findepi) [inline core/trino-main/src/main/java/io/trino/sql/planner/optimizations/ExpressionEquivalence.java:266]: left java type and right java type may be different

Comment (wendigo) [reply core/trino-main/src/main/java/io/trino/sql/planner/optimizations/ExpressionEquivalence.java:266]: Yeah. It seems that I need to implement Block sorting here as well.

File: core/trino-main/src/main/java/io/trino/sql/planner/optimizations/ExpressionEquivalence.java [hunk 3]
```diff
@@ -262,9 +266,7 @@ public int compare(RowExpression left, RowExpression right)
                     }
                 }
 
-                // value is some random type (say regex), so we just randomly choose a greater value
-                // todo: support all known type
-                return -1;
+                throw new IllegalArgumentException("Cannot compare " + leftConstant + " of type " + leftConstant.getType().getJavaType() + " and " + rightConstant + " of type " + rightConstant.getType().getJavaType());
```
Comment (findepi) [inline core/trino-main/src/main/java/io/trino/sql/planner/optimizations/ExpressionEquivalence.java:269]: - use string formatting
- printing `leftConstant.getType()` gives more information than printing `leftConstant.getType().getJavaType`, and it's shorter, so let's do that

File: core/trino-main/src/main/java/io/trino/sql/planner/optimizations/ExpressionEquivalence.java [hunk 4]
```diff
@@ -206,93 +209,89 @@ public int compare(RowExpression left, RowExpression right)
...
-
-            if (left instanceof VariableReferenceExpression leftVariableReference) {
-                VariableReferenceExpression rightVariableReference = (VariableReferenceExpression) right;
 
-                return ComparisonChain.start()
-                        .compare(leftVariableReference.getName(), rightVariableReference.getName())
-                        .compare(leftVariableReference.getType(), rightVariableReference.getType(), Comparator.comparing(Object::toString))
-                        .result();
-            }
+                    throw new IllegalArgumentException("Cannot compare %s of type %s and %s of type %s".formatted(leftConstant, leftConstant.getType(), rightConstant, rightConstant.getType()));
```
Comment (guyco33) [inline core/trino-main/src/main/java/io/trino/sql/planner/optimizations/ExpressionEquivalence.java:270]: @wendigo  It fails now on queries with Regexp
```
java.lang.IllegalArgumentException: Cannot compare ([^\\])((\\n)+) of type JoniRegExp and ([^\\])((\\n)+) of type JoniRegExp
	at io.trino.sql.planner.optimizations.ExpressionEquivalence$RowExpressionComparator.compare(ExpressionEquivalence.java:270)
	at io.trino.sql.planner.optimizations.ExpressionEquivalence$RowExpressionComparator.compare(ExpressionEquivalence.java:198)
	at io.trino.sql.planner.optimizations.ExpressionEquivalence$ListComparator.compare(ExpressionEquivalence.java:313)
	at io.trino.sql.planner.optimizations.ExpressionEquivalence$ListComparator.compare(ExpressionEquivalence.java:298)
	at com.google.common.collect.ComparisonChain$1.compare(ComparisonChain.java:114)
```

**Comment by wendigo** [comment]
@martint please take over this bug

**Pull request closed without merging**